sirepo/sim_api/jupyterhublogin.py

```diff
@@ -81,36 +85,85 @@ def init_apis(*args, **kwargs):
         ))
 
 
-def _create_user_if_not_found():
-    def __user_name(logged_in_user_name):
-        assert logged_in_user_name, 'must supply a name'
-        n = re.sub(
-            '\W+',
-            _HUB_USER_SEP,
-            # Get the local part of the email. Or in the case of another auth
-            # method (ex github) it won't have an '@' so it will just be their
-            # user name, handle, etc.
-            logged_in_user_name.split('@')[0],
-        )
-        u = JupyterhubUser.search_by(user_name=n)
-        if u or _user_dir(user_name=n).exists():
-            # The username already exists. Add a random letter to try and create
+def _create_user(github_handle=None, add_randomness=True):
+    """Create a Jupyter user and possibly migrate their data from old jupyter.
+
+    Keywords:
+      migration user: A user with data at the old jupyter
+      jupyter user: The user of new jupyter
+
+    A few interesting cases to keep in mind:
+      1. User selects to migrate but they have no old data. We have an existing
+         jupyter user who's username conflicts with the new user's github handle.
+         We should add randomness to the new user's handle.
+      2. User selects to migrate and they have old data. We should never add
+         randomness to the user's github handle because we need it to identify
+         their user dir.
+      3. User signs into sirepo under x@b.c. They migrate their data using
+         GitHub handle y. They sign into sirepo under z@b.c. They choose to
+         migrate GitHub handle y again. We should let them know that they have
+         already migrated.
+      4. A new user with no existing data selects to migrate their GitHub handle
```

Comment:
I tried to reword it. b46b58b3 . Any better? This case is very similar to case 3 except that the original user (one@any.com) gets their username from the email instead of the GitHub handle. Then the second user (two@any.com) may own that github handle (one) and truly be a different person (or the same person which would be case 3). This is subtle but I think worth remembering.

Reply:
I think [case (5)](https://github.com/radiasoft/sirepo/pull/3230#discussion_r543564226) is distinct and worth considering. The other cases are written clearly.

Reply:
I agree case 5 is worth adding. But, that doesn't answer my question if this case is now clearer?

Reply:
> The other cases are written clearly. 

:)